File: apps/api/services/render/estimate_render.py
"""Pre-export render time and cost estimation (Phase 14)."""
from __future__ import annotations

from typing import Any

from services.render.plan_render_segments import plan_render_segments

DEFAULT_SECONDS_PER_OUTPUT_MINUTE = 45.0
PARALLEL_SEGMENT_SPEEDUP = 0.65


def estimate_render(
    timeline: dict[str, Any],
    *,
    seconds_per_output_minute: float | None = None,
    usd_per_render_minute: float = 0.0,
) -> dict[str, Any]:
    """Estimate wall-clock time and optional infra cost before export."""
    fps = int(timeline.get("fps") or 30)
    duration_in_frames = int(timeline.get("durationInFrames") or 0)
    duration_seconds = (
        duration_in_frames / fps if duration_in_frames else float(timeline.get("durationSeconds") or 0)
    )
    duration_minutes = duration_seconds / 60.0 if duration_seconds else 0.0

    tracks = timeline.get("tracks") or {}
    layers = (
        len(tracks.get("motionGraphics") or [])
        + len(tracks.get("vfx") or [])
        + len(tracks.get("broll") or [])
        + len(tracks.get("captions") or [])
    )
    complexity = 1.0 + layers / 80.0
    segments = plan_render_segments(timeline)
    base = seconds_per_output_minute or DEFAULT_SECONDS_PER_OUTPUT_MINUTE

    wall_clock = duration_minutes * base * complexity
    if len(segments) > 1:
        wall_clock = (wall_clock / len(segments)) * PARALLEL_SEGMENT_SPEEDUP * len(segments)

    cost_usd = (wall_clock / 60.0) * usd_per_render_minute

    return {
        "durationSeconds": round(duration_seconds, 2),
        "segmentCount": len(segments),
        "estimatedWallClockSeconds": int(round(wall_clock)),
        "estimatedCostUsd": round(cost_usd, 4),
        "complexityScore": round(complexity, 2),
        "methodology": (
            "duration × complexity × parallel_segment_factor"
            if len(segments) > 1
            else "duration × complexity (single segment)"
        ),
        "segments": [s.to_dict() for s in segments],
    }

File: remotion-service/src/lib/render/estimateRender.ts
/**
 * Pre-export render time and cost estimation (Phase 14).
 */
import type { DirectorTimeline } from "@types/timeline";
import { planRenderSegments } from "./planRenderSegments";

export interface RenderEstimate {
  durationSeconds: number;
  segmentCount: number;
  estimatedWallClockSeconds: number;
  estimatedCostUsd: number;
  complexityScore: number;
  methodology: string;
}

export interface EstimateRenderOptions {
  /** Observed seconds of render per minute of output from past jobs. */
  secondsPerOutputMinute?: number;
  /** USD per minute of wall-clock render on paid infra (0 = local). */
  usdPerRenderMinute?: number;
}

const DEFAULT_SECONDS_PER_OUTPUT_MINUTE = 45;
const PARALLEL_SEGMENT_SPEEDUP = 0.65;

function countLayers(timeline: DirectorTimeline): number {
  const tracks = timeline.tracks ?? {};
  return (
    (tracks.motionGraphics?.length ?? 0) +
    (tracks.vfx?.length ?? 0) +
    (tracks.broll?.length ?? 0) +
    (tracks.captions?.length ?? 0)
  );
}

export function estimateRender(
  timeline: DirectorTimeline,
  options?: EstimateRenderOptions,
): RenderEstimate {
  const fps = Number(timeline.fps) || 30;
  const durationSeconds =
    Number(timeline.durationInFrames) / fps ||
    Number((timeline as { durationSeconds?: number }).durationSeconds) ||
    0;
  const durationMinutes = durationSeconds / 60;
  const layers = countLayers(timeline);
  const complexityScore = 1 + layers / 80;
  const segments = planRenderSegments(timeline);
  const baseSecondsPerMinute =
    options?.secondsPerOutputMinute ?? DEFAULT_SECONDS_PER_OUTPUT_MINUTE;

  let wallClock =
    durationMinutes * baseSecondsPerMinute * complexityScore;
  if (segments.length > 1) {
    wallClock =
      (wallClock / segments.length) * PARALLEL_SEGMENT_SPEEDUP * segments.length;
  }

  const usdPerMinute = options?.usdPerRenderMinute ?? 0;
  const estimatedCostUsd = (wallClock / 60) * usdPerMinute;

  return {
    durationSeconds,
    segmentCount: segments.length,
    estimatedWallClockSeconds: Math.round(wallClock),
    estimatedCostUsd: Math.round(estimatedCostUsd * 10000) / 10000,
    complexityScore: Math.round(complexityScore * 100) / 100,
    methodology:
      segments.length > 1
        ? "duration × complexity × parallel_segment_factor"
        : "duration × complexity (single segment)",
  };
}
